### PR TITLE
[RISCV][MC] Add support for hardcode encoding of .insn directive

### LINKED ISSUE #97498: [RISC-V] .insn directive does not support constant arg
GNU-as supports a constant value to the .insn assembler directive as of commit: https://github.com/bminor/binutils-gdb/commit/a262b82fdbf4cda3b0648b1adc32245ca3f78b7a. The RISC-V LLVM assembler currently only supports the .insn directive with in the form of `.insn <type>, <operand1>, <operand2>`. Can the form of `.insn <value>` also be supported by Clang?

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds support for the following two hardcode encodings  of .insn directive:
```
.insn <insn-length>, <value>
.insn <value>
```

See also gas's patch https://github.com/bminor/binutils-gdb/commit/a262b82fdbf4cda3b0648b1adc32245ca3f78b7a
NOTE: This patch doesn't support long instructions. 

Closes https://github.com/llvm/llvm-project/issues/97498.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 1]
```diff
@@ -1120,6 +1122,12 @@ def InsnS : DirectiveInsnS<(outs), (ins uimm7_opcode:$opcode, uimm3:$funct3,
                                         AnyReg:$rs2, AnyReg:$rs1,
                                         simm12:$imm12),
                            "$opcode, $funct3, $rs2, ${imm12}(${rs1})">;
+def Insn32 : RVInst<(outs), (ins uimm32:$value), "", "", [], InstFormatOther> {
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:1125]: Does `InstFormat` matter here?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 2]
```diff
@@ -3055,17 +3057,49 @@ bool isValidInsnFormat(StringRef Format, bool AllowC) {
...
+    if (Parser.parseIntToken(
+            Value, "expected instruction format or an integer constant"))
+      return true;
+    if (Parser.parseOptionalToken(AsmToken::Comma)) {
+      Length = Value;
+      if (Parser.parseIntToken(Value, "expected an integer constant"))
+        return true;
+    }
+
+    int64_t RealLength = (Value & 3) == 3 ? 4 : 2;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3084]: Can this be `unsigned`?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td:1125]: I don't think it should. Only a couple of the possible values are used. I think they are used to pick relocation type in the encoder. That shouldn't matter for these instructions.

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp (hunk 2 above)
Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3084]: I choose `int64_t` to avoid compiler warnings about signed-unsigned comparison.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**